Mctp bridge support

faizana-nvidia opened the pull request:
This PR aims to introduce ALLOCATE_ENDPOINT_ID message support along with MCTP Bridge endpoint into the existing peer structure.

jk-ozlabs commented:
Thanks for the contribution! I'll get to a proper review shortly.

I have some pending changes that rework a lot of the peer, link and network allocation mechanisms. That shouldn't affect your code too much, but I'll request a rebase once that is merged.

faizana-nvidia commented:
> Thanks for the contribution! I'll get to a proper review shortly.
> 
> I have some pending changes that rework a lot of the peer, link and network allocation mechanisms. That shouldn't affect your code too much, but I'll request a rebase once that is merged.

Sure no problem

jk-ozlabs reviewed (commented):
So the main design point here is how we're handling the pool allocations. It looks like your particular use-case is around static allocations, which I'll focus on here.

As I mentioned in the dbus changes, we cannot add arguments without further version-compatiblity changes. After a bit of chatting with the team, I think a better approach would be to add a new dbus call to explicitly allocate a bridge and a predefined pool (which would include the pool size). Perhaps something like:

```
AllocateBridgeStatic(addr: ay, pool_start: y, pool_size: y)
```

- where the `Set Endpoint ID` response *must* match the expected pool size.

(we would also want purely-dynamic pools to be allocated from SetupEndpoint and friends, but that would result in a dynamic pool allocation. This dynamic pool would be defined either by a toml config option, or via a new TMBO dbus interface. However, we can handle those later, I think)

Would that work?

jk-ozlabs commented:
In general, can you add a bit more of an explanation / rationale as part of your commit messages, instead of just log output? There is some good guidance for commit messages up in the "Patch formatting and changelogs" section of https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/5.Posting.rst

faizana-nvidia commented on src/mctp-control-spec.h:
```diff
@@ -175,6 +175,29 @@ struct mctp_ctrl_resp_resolve_endpoint_id {
+
+struct mctp_ctrl_cmd_alloc_eid {
+	struct mctp_ctrl_msg_hdr ctrl_hdr;
+	mctp_ctrl_cmd_alloc_eid_op alloc_eid_op : 2;
+	uint8_t : 6;
```
Thank you for your comments, your ask makes sense to me, have tried to go via similar approach as done in mctp_ctrl_cmd_set_eid where bitfield was used. Nonetheless I will change this and rely more on bitmask/shift operation

faizana-nvidia commented on docs/mctpd.md:
```diff
@@ -123,16 +123,18 @@ Similar to SetupEndpoint, but will always assign an EID rather than querying for
 `mctpd`.
 
-#### `.AssignEndpointStatic`: `ayy` → `yisb`
+#### `.AssignEndpointStatic`: `ayyy` → `yisb`
```
I feel like there's a misunderstanding, the new argument was to take in pool_start-eid while the pool size itself would have been filled based on whatever bridge mentions as its pool size.

About this breaking the existing dbus api, I totally agree with you on this, based on your suggestion I agree may be having a new dbus method for AllocateBridgeStatic would be more convenient and stable way of doing this.

But I do have few questions on the suggested signature
AllocateBridgeStatic(addr: ay, pool_start: y, pool_size: y)

Q1) Do we really need pool_size as an argument here? (I get that your next point on checking this with the response of Set Endpoint ID command, but I don't get if its necessary)

_where the Set Endpoint ID response must match the expected pool size._
Q2) If we go with seperate dbus method for Allocating EIDs, we would need own bridge EID as an argument as well.
AllocateBridgeStatic(addr: ay, bridge_eid: y, pool_start: y)

But this would again lead us to split the method (Static and Dynamic) to cover dynamic EID allocation as well.
I presume this is what you meant from 

_(we would also want purely-dynamic pools to be allocated from SetupEndpoint and friends, but that would result in a dynamic pool allocation. This dynamic pool would be defined either by a toml config option, or via a new TMBO dbus interface. However, we can handle those later, I think)_

i.e for dynamic pool allocation we would prefer SetupEndpoint and AssignEndpoint method? Please correct me if I misunderstood you.

jk-ozlabs replied:
> Q2) If we go with seperate dbus method for Allocating EIDs, we would need own bridge EID as an argument as well.

Yes, good point. I had intended to include this, but missed it in the prototype.

> Q1) Do we really need pool_size as an argument here?

Yes, I think so.

With the static strategy that you're proposing here, the general interaction model will be that some external process (say, OpenBMC's Entity Manager) is taking responsibility for controlling the EID space. With the current approach (of just specifying the start), there is split responsibility: the bridge itself now controls part of the allocation scheme. To provide a consistent approach to the allocations, we would need control of the pool size so that we're not at risk of creating an overlapping pool with some other (future) bridge pool allocation.

We probably have a choice about what to do in case those pool requests do not match (ie, the call argument vs. the Set Endpoint ID response) - either refusing the pool allocation entirely, or truncating the pool to the smaller of those values. I think the latter would be reasonable.

> But this would again lead us to split the method (Static and Dynamic) to cover dynamic EID allocation as well.

We don't really need to split it - the existing calls (SetupEndpoint, AssignEndpoint and possibly AssignEndpointStatic) would just gain support for the pool allocation as well - depending on what the device requests from the Set Endpoint ID response. In those cases, if a pool is requested, it would be completely dynamic, and we would serve that from a per-whole-network EID pool (configurable, via the toml file).

How does that sound?

faizana-nvidia replied:
Yeah sounds feasible and reasonable to me (should have thought about case when ask won't be of utilizing full bridge size but a short chunk of it, giving better control to upper applications like dbus sensors etc.). So we'll have a poll from start eid to pool size asked, to check presence of device on the network. But in case we are going with dynamic eid allocation, since pool size will be relied on what whatever bridge has as capacity, we'll utilizing all the eids from the pool.

Whatever pool size we ask either it would be more than bridge capability in which case, it would assign only to its max capability thus truncating the value and us getting notified via response of Allocate endpoint ID. On contrary, if asks is less than max capability then bridge would do it since we'll be updating the request with required size.

faizana-nvidia commented on tests/conftest.py:
```diff
@@ -291,6 +292,10 @@ async def handle_mctp_control(self, sock, addr, data):
                 (op, eid) = data[2:]
                 self.eid = eid
                 data = bytes(hdr + [0x00, 0x00, self.eid, 0x00])
+                if (self.pool_size):
+                    data = bytes(hdr + [0x00, 0x01, self.eid, self.pool_size])
+                else:
+                    data = bytes(hdr + [0x00, 0x00, self.eid, 0x00])
```
Acknowledge it, will update it

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -2173,6 +2184,11 @@ static int method_assign_endpoint_static(sd_bus_message *call, void *data,
 	dfree(peer_path);
 
+	/* MCTP Bridge Support */
+	if(peer->pool_size > 0) {
```
my bad, since we are on this, what clang-format are we following here? Any steps on how I could run it locally. Would come in handy.

jk-ozlabs replied:
We don't have a full .clang-format definition, but we're roughly doing kernel format here. If it helps, I can put an actual config together, but that then we'd need a lot of churn in applying it.

So, that's probably better something done just before release, to avoid to much merge collisions.

Or, if you're OK with a bit of churn, I can do that beforehand?

faizana-nvidia replied:
I'm sold for before release part :)

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -4494,6 +4509,160 @@ static void setup_config_defaults(ctx *ctx)
+		peer->pool_size = allocated_pool_size;
+		peer->pool_start = allocated_pool_start;
+
+		//delay for settling MCTP Bridge after allocation of downstream eid
```
Let me highlight the issue better, it's not that Allocate Endpoint ID is getting delayed (we are sending it as part of sequence right after bridge eid is assigned), but post allocate endpoint id, referring to my setup, I see no response from the downstream EIDs, basically timeout for GET_MESSG_TYPE and GET_UUID mctp commands on downstream eids.

On topic of retry mechanism, in my setup based on observational data, find that it takes around 4 sec for all eids to be in proper state after receiving ALLOCATE_EID but this also depends on the bus topology and how many downstream endpoints are there.

That being said it could easily be something which could become a common issue for Bridge's owing to their network congestion, resource congestion etc.
A retry mechanism would then be needed as general part of design for messages like GET_MESSG_TYPE and GET_UUID, which is currently missing implementation.

Would like to explore any other suggestion beyond retry mechanism, if possible :)

jk-ozlabs replied:
> Let me highlight the issue better, it's not that Allocate Endpoint ID is getting delayed (we are sending it as part of sequence right after bridge eid is assigned), but post allocate endpoint id, referring to my setup, I see no response from the downstream EIDs, basically timeout for GET_MESSG_TYPE and GET_UUID mctp commands on downstream eids.

ah, gotchya!

Yeah, this isn't a trivial problem, and I've been thinking of potential approaches for a while.

My current leaning is that once an endpoint pool is assigned, mctpd would perform some sort of polling to determine which of the pool's EIDs are active (ie, have been assigned to a device, and that device is responding). 

We could potentially do this via a Get Routing Table Entries command to the bridge, but that's a bit of a mis-use of the command, and isn't necessarily returning the data we want (the bridge may have routing information to a device, but the device may not be present).

So, I think the most reliable approach is that the pool's EIDs are periodically polled using a Get Endpoint ID command - which is the spec-recommended method of detecting endpoint presence, and is what we use for the existing recovery process. We may be able to share some code with the recover infrastructure for this too.

We would schedule those polls when the pool is allocated, and then repeat until we detect a device - at which point we would perform full enumeration. If no device is ever present, mctpd would end up polling forever, but I don't think there's an alternative there.

santoshpuranik replied:
How about doing a `GetRoutingTableEntries`, the bridge should return `ERROR_NOT_READY` if it is still assigning EIDs downstream from the previous `AllocateEndpointID`. Once we get a successful response, we can start querying EIDs downstream for UUIDs and message types.

jk-ozlabs replied:
> the bridge should return ERROR_NOT_READY if it is still assigning EIDs downstream from the previous AllocateEndpointID

that's not necessarily what the bridge will do though - there's no concept of the routing state being "complete" at any point (eg, for a downstream port that may have an enpoint hotplugged at some point in the future (or perhaps never!))

For that reason, I don't think we'll be able to avoid polling.

faizana-nvidia replied:
one downside of this I feel (if we don't go via GetRoutingTable implementation (have a change list ready but nevermind)) is we'll be ending up utilizing the eids from the pool (specially for Dynamic eid allocation based on bridge pool size) which is a bit concerning if the network becomes complex.

Take RoutingTable approach makes us definitive about what eids needs to be consumed from the pool (minimizing the utilization, based on available device downstream).

faizana-nvidia replied:
> Not sure I understand what you mean by "utilizing the eids" here, could you elaborate?

Sorry I should have explained better. May be I'm overthinking it here but here's what I meant

For instance, if we have a network comprising of multiple Bridge's where we have to assign all Dynamic Eids to the downstream, as per our current assumption, we'll have to keep aside all Eids which bridge needs based on its pool size (Bridge eid : 8, pool size :15 thus 9- 23 is consumed from our Dynamic Pool even though there's possibility of not having devices downstream for 9-23 Eids, since we carve eids from our maintained pool via (code below) and then add_peer(consumed here) later update setup_peer()

`/* Find an unused EID */
		for (e = eid_alloc_min; e <= eid_alloc_max; e++) {
			if (n->peeridx[e] == -1) {
				rc = add_peer(ctx, dest, e, net, &peer);
				if (rc < 0)
					return rc;
				break;
			}
		}`
this could end up consuming EIDs for the device which may not be there.

Also to mention based on the allocation of EIDs needs to be contiguous, this would create another issue if there happens to be other Bridges which has been allocated EID statically which kind of makes find a contiguous set of EIDs for the downstream endpoints of Dynamic Allocated Bridge difficult and where need of next contiguous set of EIDs rises.

**Spec : DSP0236 Version: 1.3.0**

**9.1.9 Consolidating routing table entries**
_MCTP requires that when an EID pool is allocated to a device, the range of EIDs is contiguous and
 follows the EID for the bridge itself. Thus, a bridge can elect to consolidate routing table information into
 one entry when it recognizes that it has received an EID or EID range that is contiguous with an existing
entry for the same physical address and bus. (The reason that EID allocation and routing information
 updates are not done as one range using the same command is because of the possibility that a device
may have already received an allocation from a different bus owner.)_

In such contingency with limited  EIDs availability, keeping EIDs for non available device is what I fear. Though I agree with your point too Routing Table may not be the best approach to detect Endpoint presence in the network especially if the endpoint is hot pluggable.

> My concern with using Get Routing Table Entries is that we're on the edges of what the spec describes this command for, whereas Get Endpoint ID is mentioned as a mechanism for this. An endpoint being described in Get Routing Table Entries does not mean it is available, so we would have to fall back to querying/polling it with Get Endpoint ID _anyway_.

 But then again in such case Spec tells us about commands like Discovery Notify which should be initiated by the Bridge to notify any change in current network topology under the bridge. Please feel free to correct me in my assumptions :)

jk-ozlabs replied:
> we'll have to keep aside all Eids which bridge needs based on its pool size (Bridge eid : 8, pool size :15 thus 9- 23 is consumed from our Dynamic Pool even though there's possibility of not having devices downstream for 9-23 Eids
[...]
> this could end up consuming EIDs for the device which may not be there.

Yes, absolutely. If we have done an Allocate Endpoint IDs that gives a range of EIDs to a bridge, then mctpd no longer has ownership of any of those EIDs, and cannot hand out any subset of that range elsewhere.

> In such contingency with limited EIDs availability, keeping EIDs for non available device is what I fear.

We have no choice about that - if we allocate a range to a bridge, the entire range belongs to the bridge, regardless of whether the EIDs in that range are then allocated to downstream devices, or whether those downstream devices are active.

> But then again in such case Spec tells us about commands like Discovery Notify which should be initiated by the Bridge to notify any change in current network topology under the bridge.

That's not my interpretation of the spec for Discovery Notify; this is more to announce endpoint presence, and is only used on specific transport types (ie, those that do not have their own presence detection mechanism).

faizana-nvidia replied:
understood, Thank you so much for clearing this out, I'm gonna integrate the polling logic via GET_ENDPOINT_ID similar to recovery.

Since we dont have kernel implementation for using the bridge as a gateway for that range can we go with sub-optimal way of thing ? i.e adding routes and neighbours to these discovered device via Bridge phys address till the time your changes are getting ready?

faizana-nvidia commented on tests/conftest.py:
```diff
@@ -1031,8 +1037,8 @@ def name_owner_changed(name, new_owner, old_owner):
-Contains one interface (lladdr 0x10, local EID 8), and one endpoint (lladdr
-0x1d), that reports support for MCTP control and PLDM.
+Contains two interface (lladdr 0x10, local EID 8), (lladdr 0x11, local EID 10) with one endpoint (lladdr
+0x1d), and MCTP bridge (lladdr 0x1e, pool size 2), that reports support for MCTP control and PLDM.
```
I see, so better to update the current interface (lladdr 0x10, local EID 8) with pool size and update pool size numbered eids to the network simulating a bridge rather than creating a new one?

jk-ozlabs replied:
Yep! Just update the interface/endpoints/etc in the individual test case.

jk-ozlabs commented:
We'll also need to consider the routing setup for bridged endpoints. Ideally we would:

 1. create a route for the bridge itself, plus a neighbour entry with the appropriate physical address data
 2. create a range route for the allocated endpoint pool, using the bridge as a gateway for that range (ie, no neighbour entry)

the issue is that there is no kernel support for (2) at present: we need some kernel changes to implement gateway routes. It *is* possible to create "somewhat-fake" routes for those endpoints, using a neighbour table entry for each (bridged) peer that uses the bridge phys address, but that's a bit suboptimal. I'd prefer not to encode that hack into mctpd if possible.

I do have a todo for the kernel changes necessary for that, sounds like I should get onto it!

santoshpuranik commented:
> We'll also need to consider the routing setup for bridged endpoints. Ideally we would:
> 
>     1. create a route for the bridge itself, plus a neighbour entry with the appropriate physical address data
> 
>     2. create a range route for the allocated endpoint pool, using the bridge as a gateway for that range (ie, no neighbour entry)
> 
> 
> the issue is that there is no kernel support for (2) at present: we need some kernel changes to implement gateway routes. It _is_ possible to create "somewhat-fake" routes for those endpoints, using a neighbour table entry for each (bridged) peer that uses the bridge phys address, but that's a bit suboptimal. I'd prefer not to encode that hack into mctpd if possible.
> 
> I do have a todo for the kernel changes necessary for that, sounds like I should get onto it!

IIUC, 1 is what we can achieve with the tools we have today, right? For ex: add route to the bridge itself and then `mctp route add <downstream eid> via <bridge net if>`, essentially adding a neighbour table entry? Would this not continue to work as from TMBO point-of-view all packets go via the bridge route.

When you say sub-optimal, are you referring to the neighbour lookup that happens in `net/mctp/route.c`? Noob question, how does a gateway impl make that faster?

When is the gateway support in kernel for MCTP nets planned? We can help if you have a design in mind.

jk-ozlabs commented:
Hi Santosh,

> IIUC, 1 is what we can achieve with the tools we have today, right?

Yes, but it requires a lot of workaround to set up.

> For ex: add route to the bridge itself and then `mctp route add <downstream eid> via <bridge net if>`, essentially adding a neighbour table entry?

That isn't adding a neighbour table entry though; just a route. USB is a little different in that there are no neighbour table entries required, because there is no physical addressing.

For a bridge, using this scheme would require:

 1. adding the route to the bridge EID
 2. adding the neighbour entry for the bridge EID
 3. adding individual routes for each EID in the EID pool
 4. adding individual fake neighbour table entries for each EID in the EID pool, which would (incorrectly) represent that the EID has a specific physical address (ie., that of the bridge)

(for USB, we don't need (2) or (4), but that's purely a property of the transport type. We would need those to be supported in mctpd to allow other transport types like i2c).

This would work, but it's messy.

> When you say sub-optimal, are you referring to the neighbour lookup that happens in `net/mctp/route.c`?

No, the neighbour lookups happen in `net/mctp/neigh.c`.

> Noob question, how does a gateway impl make that faster?

Not so much faster, more tidier. With a gateway route, we would require:

 1. adding the route to the bridge EID
 2. adding the neighbour entry for the bridge EID
 3. adding one range route for the entire EID pool, referencing the bridge EID as the gateway

No fake neighbour table entries are required - since the kernel just looks up the gateway physical address from the gateway's neighbour table entry.

> When is the gateway support in kernel for MCTP nets planned?

I have it done - will push a development branch shortly.

jk-ozlabs commented:
> I have it done - will push a development branch shortly.

https://github.com/CodeConstruct/linux/tree/dev/forwarding

santoshpuranik commented:
Hi Jeremy,

Thank you for the detailed response.

> Hi Santosh,
> 
> > IIUC, 1 is what we can achieve with the tools we have today, right?
> 
> Yes, but it requires a lot of workaround to set up.
> 
> > For ex: add route to the bridge itself and then `mctp route add <downstream eid> via <bridge net if>`, essentially adding a neighbour table entry?
> 
> That isn't adding a neighbour table entry though; just a route. USB is a little different in that there are no neighbour table entries required, because there is no physical addressing.
> 
> For a bridge, using this scheme would require:
> 
>     1. adding the route to the bridge EID
> 
>     2. adding the neighbour entry for the bridge EID
> 
>     3. adding individual routes for each EID in the EID pool
> 
>     4. adding individual fake neighbour table entries for each EID in the EID pool, which would (incorrectly) represent that the EID has a specific physical address (ie., that of the bridge)

Ack, I see something like I2C would need a PHY address.

> > When you say sub-optimal, are you referring to the neighbour lookup that happens in `net/mctp/route.c`?
> 
> No, the neighbour lookups happen in `net/mctp/neigh.c`.
> 

Ack. I should have said the neigh_lookup call that happens in route.c!

> > Noob question, how does a gateway impl make that faster?
> 
> Not so much faster, more tidier. With a gateway route, we would require:
> 
>     1. adding the route to the bridge EID
> 
>     2. adding the neighbour entry for the bridge EID
> 
>     3. adding one range route for the entire EID pool, referencing the bridge EID as the gateway
> 
> 
> No fake neighbour table entries are required - since the kernel just looks up the gateway physical address from the gateway's neighbour table entry.
>

Thank you, that does seem cleaner.

jk-ozlabs commented:
And for the userspace changes, my `dev/gateway` branch here:

https://github.com/CodeConstruct/mctp/tree/dev/gateway

santoshpuranik commented:
@jk-ozlabs : I think we agree that mctpd has to poll all allocated endpoints with a Get Endpoint ID periodically. I think the first thing we'd need to enable in order to do that is to make MCTP requests and responses asynchronous. Do you have a design in mind to make MCTP requests async (like via a request queue per allocated endpoint)?

jk-ozlabs commented:
> I think we agree that mctpd has to poll all allocated endpoints with a Get Endpoint ID periodically

Just as a clarification -  not _all_ endpoints, but EIDs within allocated endpoint ranges, which have not yet been enumerated. And this is assuming we expect mctpd to automatically enumerate those bridged devices. I think the latter is reasonable, but we don't have a specific design point around that yet.

With that in mind, yes, we probably want to make that async, as those requests are likely to not have a response, and therefore we're at worst-case waiting time.

In terms of design: we probably don't want a `struct peer` to be created for those endpoints, as we they don't strictly exist as proper peers at that stage. I think a minimal-impact approach may be to keep a set of the allocated (but not-yet-enumerated) ranges, and periodically send the Get Endpoint ID requests.

We don't necessarily need to keep much state for that polling mechanism (ie, between request and response) - receiving a Get Endpoint ID response for anything in that range would trigger the enumeration process.

santoshpuranik commented:
> No, and we don't do that with directly-attached endpoints either.

So we have a case where we will have to call allocate endpoint ID on the bridge device when not all of its downstream devices are available. In such a case, how do you think we can determine when those downstream EIDs become available unless we poll?

amboar commented:
> .. and speaking of `Recover`, we might need to revisit how we handle that for bridged endpoints, as a MCTP-level recovery operation probably isn't applicable as a directly-attached device (in the same manner, at least). CC @amboar.

It will need some rework as currently it assumes the peer is a neighbour and uses physical addressing for `Get Endpoint ID`. We still want a mechanism to establish the loss of a non-neighbour peer though. I think `Recover` is fine for that. We need to use some message for polling, and despite the absurdity I think `Get Endpoint ID` is also fine for that, just we can't use physical addressing if the peer is behind a bridge. The observed behaviour of `Recover` would be the same - if the peer is responsive then the D-Bus object remains exposed, or if it's unresponsive then the object is removed. The difference between a peer being unresponsive as opposed to not having yet been assigned an address cannot be determined across the bridge, so in that case we skip the substance of the recovery operation (EID re-assignment). That's a responsibility of the bridge node anyway.

amboar commented:
Yep, that sounds sensible.

faizana-nvidia commented:
Thank you all for taking out time to look into the PR,

I've addressed to the asked comments on previous commit, added new commit for MCTP Bridge design doc, need to push Polling mechanism now

jk-ozlabs commented:
Thanks for the updates! A couple of comments:

1) We don't really do design proposal docs as file in the repo; it's great to see your recap of the discussion points from this PR, but there's no need for that format to be long-lived in the repo itself. I would suggest turning this into a user-consumable document describing how things work according to your new implementation. Any dbus API changes belong in the mctpd.md document.

2) Before implementing this new dbus API, we would need some confirmation that non-contiguous pool allocations are permissible. I have raised an issue with the PMCI WG, (#1540, if you have access), and would like at least some indication that the requirement can be relaxed before we commit to the separate pool ranges.

3) In order to reduce the upfront work, you may want to skip the endpoint polling for the initial PR; the changes will still be useful in that `au.com.codeconstruct.MCTP.Network1.LearnEndpoint` can be used to enumerate downstream devices manually (once the pool is allocated, and we can route to those endpoints).

4) You have a couple of cases where you add something in an initial patch, then re-work it in the follow-up patch. This makes review overly complicated.

5) Super minor, but the formatting of introduced changes is inconsistent. Given there's still some work to do before this series is ready, I will apply the tree-wide reformat shortly, and add a .clang-format.

faizana-nvidia commented:
> Thanks for the updates! A couple of comments:
> 
> 1. We don't really do design proposal docs as file in the repo; it's great to see your recap of the discussion points from this PR, but there's no need for that format to be long-lived in the repo itself. I would suggest turning this into a user-consumable document describing how things work according to your new implementation. Any dbus API changes belong in the mctpd.md document.
> 2. Before implementing this new dbus API, we would need some confirmation that non-contiguous pool allocations are permissible. I have raised an issue with the PMCI WG, (#1540, if you have access), and would like at least some indication that the requirement can be relaxed before we commit to the separate pool ranges.
> 3. In order to reduce the upfront work, you may want to skip the endpoint polling for the initial PR; the changes will still be useful in that `au.com.codeconstruct.MCTP.Network1.LearnEndpoint` can be used to enumerate downstream devices manually (once the pool is allocated, and we can route to those endpoints).
> 4. You have a couple of cases where you add something in an initial patch, then re-work it in the follow-up patch. This makes review overly complicated.
> 5. Super minor, but the formatting of introduced changes is inconsistent. Given there's still some work to do before this series is ready, I will apply the tree-wide reformat shortly, and add a .clang-format.

Hello Jeremy

Thank you for looking over the commits, based on your comment #1 I have removed the new .md file which captured MCTP Bridge support details on PR and updated the existing mctpd.md file with new information about dbus api AssignBridgeStatic. Regarding user consumable document, I'm not much sure what this could be, if you could let me know what this document should be I can create one and update the PR.

I recently got the permission  for PMCI WG, have glanced over what was stated on issue #1540, basically the Idea is to split the BusOwner EID pool and segregate a chunk of eids for Bridge's downstream pool on the higher end of Busowner pool while keeping lower end for non-bridge devices. This would be helpful for Dynamic EID assignment of downstream pool devices incase multiple Bridge's are there under same network.

My current implementation involves finding of contiguous eid chunk of min(requested pool size, bridge's pool size capability) from available BusOwner's pool but we begin looking from Asked pool_start (Static) or from next to Bridge's EID (Dynamic) and we look till we get right sized chunk and mark that eid as pool_start. I did based this from the same line of spec for which you raised the issue.


>_In order to reduce the upfront work, you may want to skip the endpoint polling for the initial PR; the changes will still be useful in that au.com.codeconstruct.MCTP.Network1.LearnEndpoint can be used to enumerate downstream devices manually (once the pool is allocated, and we can route to those endpoints)._

**I can create a new PR for Endpoint polling if thats what you mean and skip for this PR**. Also for adding route for the downstream endpoint to Bridge, we would need your implementation implementation to be merged for both linux kernel and mctpd. Internally I've tested my polling logic with your pulled changes, but for this PR I haven't picked them up so discovery of downstream EID via LearnEndpoint would probably not be possible with only this PR


> 4. You have a couple of cases where you add something in an initial patch, then re-work it in the follow-up patch. This makes review overly complicated.
> 5. Super minor, but the formatting of introduced changes is inconsistent. Given there's still some work to do before this series is ready, I will apply the tree-wide reformat shortly, and add a .clang-format.

I've updated the patch set now for easier review, hope it helps, let me know if I can do anything else to further ease the review. Thanks for your .clang format, once that is pushed I would reply those onto my change

faizana-nvidia commented:
> Sounds good. I see you have created another PR that involves both the allocation scheme and the polling though. I would suggest splitting this, or just including the allocation mechanism in this PR.

Sure I can chery-pick the commit [955d1b4c](https://github.com/CodeConstruct/mctp/pull/85/commits/955d1b4c08fe89b2cc02620820125ce4e0a6b5ce#diff-222eb3ffb54581db6e3d72a6efa39bf25a8145a7f8eab9b7f4474ea184675da9R3790) (from [PR 85)](https://github.com/CodeConstruct/mctp/pull/85) onto the current PR 71. 

> In our case, we may still want to do this sequentially (ie, the pools still following the bridge EID), but that may not always be possible (for example, if that contiguous range would conflict with an existing allocation), so we need a facility to separate the allocations.

right for that case I have added checks while finding out the right contiguous pool start to see if the chunk has any EID which is associated to any other bridge [link to other PR 85](https://github.com/CodeConstruct/mctp/pull/85/commits/955d1b4c08fe89b2cc02620820125ce4e0a6b5ce#diff-222eb3ffb54581db6e3d72a6efa39bf25a8145a7f8eab9b7f4474ea184675da9R4612). This helps to find the conflict and we skip on to other pool_start.

jk-ozlabs commented:
> Now the question here is (based on what you said and is defined in specs _the current DSP0236 wording does not allow us flexibility over the start of the pool anyway..._) should we not extend this capability of picking up the start of pool via dbus method? or are we to assume the very next EID of the bridge's own EID to be the start of the pool because it needs to be contiguous with bridge's own eid?

We need compatibility with this requirement at least. So, given that you mention:

> In our platform we really need to have start of pool be deterministic.

\- so we have to use a separate dbus call that provides the pool start (and possibly a max pool size) - like you have already proposed with `AssignBridgeStatic`. So all good there. We will need to commit to the pool range specified in that dbus call.

Given the contiguous requirement that currently exists, that dbus call will (currently) need to immediately fail if the pool start is not the static bridge EID + 1. It will also need to fail if the entire proposed pool is not currently free.

So, given we have a-priori knowledge of the EID pool ('cos we're doing everything static in this case) - for the behaviour of `AssignBridgeStatic`:

 - fail on pool_start != eid + 1
 - fail if  any EID in the range [eid, pool_start + pool_end] is not currently free
 - immediately (internally) allocate that EID range to the bridge peer
 - perform Set Endpoint Id
   - if the Set Endpoint Id response indicates a pool size *smaller* than specified in the dbus call, we may want to shrink the pool, but that can come as a later patch
 - set up a single-EID route to the bridge EID
 - perform Allocate Endpoint IDs
   - again, if the Allocate Endpoint Ids response indicates a pool size *smaller* than specified in the dbus call, we may want to shrink the pool, but that can come as a later patch. It's a bit odd that the bridge would respond with a smaller pool size than it originally requested in the Set Endpoint Id response, so I assume this is rare.
 - set up a EID-range route to the bridge pool

> In both way actual pool start is something would be considered based on response of AllocateEID byte 4. So what I'm trying to get at is, we can't really consider EIDs reserved till we have the knowledge of its start and allocated_pool size.

We *have to* reserve those EIDs until we have confirmation that they are unused - from the Set Endpoint Id call (where we commit to a bridge EID) onwards. Otherwise the allocation may become invalid during the assignment process.

At the moment, this isn't really significant, as the endpoint setup procedure is synchronous (for example, the Assign Endpoint Ids command is processed immediately after Set Endpoint Id, and no allocations may have occurred between them). However, if we make it async in future, we need the reservations to be valid at any point in the process, so I suggest we get this right to start with.

> Just curious about having it all, why can't we keep both?

Because it may be an unexpected change in behaviour for the existing calls. I'll come up with a plan for those, but it shouldn't affect your implementation here.

santoshpuranik commented:
@jk-ozlabs : A couple of questions:

Is the comment Yuval made on https://github.com/DMTF/PMCI-WG/issues/1540 not sufficient for us to drop the need for the pool to start immediately after the bridge EID? Would you need to see that as a spec. update?

Given that we do not have consensus in the OpenBMC world yet on how to do static EIDs (that is, how to write EM configs for these), do you think we should do dynamic EID assignments instead (no new D-Bus API for static). That way we have something that is immediately usable. IOW, we do option 1) from your comment here: https://github.com/CodeConstruct/mctp/pull/71#issuecomment-3051859865

jk-ozlabs commented:
> Is the comment Yuval made on [DMTF/PMCI-WG#1540](https://github.com/DMTF/PMCI-WG/issues/1540) not sufficient for us to drop the need for the pool to start immediately after the bridge EID? Would you need to see that as a spec. update?

No, this is not sufficient; even if a new spec was issued, we would still need compatibility with pre-update hardware implementations.

(I'd be fine with a config setting that relaxes this requirement, but we would need to be spec-compliant by default)

> Given that we do not have consensus in the OpenBMC world yet on how to do static EIDs (that is, how to write EM configs for these), do you think we should do dynamic EID assignments instead (no new D-Bus API for static). That way we have something that is immediately usable.

I'm basing the requirements on the mention that the pool ranges need to be deterministic (per @faizana-nvidia ); that won't be guaranteed if we integrate the bridge pool allocations into the existing Setup / Assign / Learn interfaces.

Or is that not necessary?

Even without the EM updates, it's still entirely possible to use the proposed AssignBridgeStatic interface; we'd just have a different facility to invoke it.

jk-ozlabs commented:
> What do you think is needed to allow both?

Mainly a facility to ensure that the static and dynamic allocations can coexist. My current plan for this is a way to configure an explicit range of EIDs that mctpd can dynamically allocate, and then the rest are reserved for static allocation.

> Currently, doing dynamic seems like something that can be pushed through easier.

OK. On the mctpd side, static is certainly easier, as we don't need to worry about calculating any allocations :)

But for dynamic, I would think something like:

 1. As with non-bridge endpoints, AssignEndpoint[^1] is called
 2. mctpd finds a suitable EID allocation (bridge EID + bridge pool) for the endpoint:
    - If we're intending to conform to the contiguous EID range requirement, we would need to find the largest available EID range from the dynamic EID pool, and assign the lowest EID from that range. (We may want to decide on an arbitrary maximum pool size here).
    - If we're not, we can chose any available EID from the dynamic pool, and defer the bridge pool allocation to later, when we know the size
  3. immediately (internally) allocate that entire EID range to the new peer (even though we don't know whether we're assigning a single EID endpoint or a bridge, due to the contiguous EID requirement)
  4. perform Set Endpoint ID
     - if the Set Endpoint ID response indicates no pool request, we're done, release all remaining EIDs
     - if the Set Endpoint ID response indicates a pool request, ensure that it fits the allocation, and shrink the allocation if the pool request is smaller
  5. perform Allocate Endpoint IDs

We would likely want a new dbus interface on endpoints which report as bridges, with properties indicating the pool range.

Also, this assumes that the AssignEndpoint caller does not need to handle bridge pools specially; if so, we'll likely need another approach.

Then, as a follow-up:

 - After the Allocate Endpoint IDs response, start polling the bridge range, to enumerate subordinate endpoints as they appear

[^1]: note that only AssignEndpoint would work as the entry-point here; LearnEndpoint will not involve a Set Endpoint ID command, and SetupEndpoint may not involve one. This means we'll need to be particularly mindful of resets; they can not preserve remote peer EID states like SetupEndpoint would.

jk-ozlabs reviewed (commented):
In general, it looks like we're converging on a good approach here.

A few comments inline, and one overall: I'm not sure that the addition of the `reserved_eid_set` is for, when we already have the pool information in the peers list. Can we implement the allocation without duplicating this data?

We'll definitely need more tests though, so check cases of allocation failures, and cases where we need to constrain the pool requests from the peer.

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -1662,7 +1669,7 @@ static int peer_set_mtu(struct ctx *ctx, struct peer *peer, uint32_t mtu)
 static int endpoint_assign_eid(struct ctx *ctx, sd_bus_error *berr,
 			       const dest_phys *dest, struct peer **ret_peer,
-			       mctp_eid_t static_eid)
+			       mctp_eid_t static_eid, bool check_bridge)
```
Wanted to enforce setting up of Bridge only with AssignEndpoint dbus-call. Since endpoint_assign_eid() is the core logic common to all other method such as SetupEndpoint or AssignEndpointStatic, thus the check.

Had to modify EID check logic here endpoint_assign_eid()

faizana-nvidia replied:
renamed the varaible to avoid confusion.

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -45,6 +45,8 @@
 #define CC_MCTP_DBUS_IFACE_BUSOWNER "au.com.codeconstruct.MCTP.BusOwner1"
 #define CC_MCTP_DBUS_IFACE_ENDPOINT "au.com.codeconstruct.MCTP.Endpoint1"
+#define CC_MCTP_DBUS_IFACE_ENDPOINT_TYPE \
+	"au.com.codeconstruct.MCTP.EndpointType1"
```
okay I did think of this first and pushed the change, but later decided to cover all aspect of endpoints. Sure what you said makes sense too. I'll update it back

faizana-nvidia commented:
> In general, it looks like we're converging on a good approach here.
> 
> A few comments inline, and one overall: I'm not sure that the addition of the `reserved_eid_set` is for, when we already have the pool information in the peers list. Can we implement the allocation without duplicating this data?
> 
It could be implemented, we'll just have to fetch eids presence, in another bridge peer quite a lot many times (felt like less efficient at the moment)

But checking via peer would be cleaner and less maintainable way, while BITMAP would be more efficient owing to frequent check we need to do while looking for contiguous available eids per bridge.

I'm okay with removing it. Let me know.

jk-ozlabs commented:
Not sure I fully understand your response there:

>  we'll just have to fetch eids presence

what do you mean by "fetch eids presence"? where is this fetching from?

> But checking via peer would be cleaner and less maintainable way

why less maintainable? it would be one fewer data structure to keep in sync.

I would suggest you add the EID allocator as a new function, that:

- creates a temporary bitmap for the available EIDs
- walks the peers list for this network, updating the bitmap according to each peer's ->eid and ->pool_start & pool_size
- finds the required range from the bitmap

there's no real need for the bitmap to be persistent; we're only dealing with a fixed number of peers here, and only need to do this when performing the initial allocation (so once per discovered peer).

This way, all of the allocation logic is in a single function, and we cannot have a state where the map is out of sync with the actual peers list.

faizana-nvidia commented on conf/mctpd.conf:
```diff
@@ -4,6 +4,7 @@ mode = "bus-owner"
 # MCTP protocol configuration. Used for both endpoint and bus-owner modes.
 [mctp]
 message_timeout_ms = 30
+MAX_POOL_SIZE = 15
```
Ack, will update

faizana-nvidia commented:
> Not sure I fully understand your response there:
> 
> > we'll just have to fetch eids presence
> 
> what do you mean by "fetch eids presence"? where is this fetching from?

By fetching i mean, iterating over peers array to find if the eid is used or not, we also now need to check if eid is falling within range of any peer's allocated pool (peer->pool_start  to  peer->pool_start + peer->pool_size).

But I'll see if we can avoid bitmap. If not then will implement via one function way
> 
> > But checking via peer would be cleaner and less maintainable way
> 
This is meant for non bitmap approach
> why less maintainable? it would be one fewer data structure to keep in sync.
> 
Yes I agree
> I would suggest you add the EID allocator as a new function, that:
> 
> * creates a temporary bitmap for the available EIDs
> * walks the peers list for this network, updating the bitmap according to each peer's ->eid and ->pool_start & pool_size
> * finds the required range from the bitmap
> 
> there's no real need for the bitmap to be persistent; we're only dealing with a fixed number of peers here, and only need to do this when performing the initial allocation (so once per discovered peer).
> 
> This way, all of the allocation logic is in a single function, and we cannot have a state where the map is out of sync with the actual peers list.

mkj commented on src/mctpd.c:
```diff
@@ -268,6 +288,83 @@ __attribute__((format(printf, 1, 2))) static void bug_warn(const char *fmt, ...)
+}
+
+// Clear the entire set
+void eid_set_clear(reserved_eid_set *set)
```
Duplicate of `eid_set_init`

mkj commented on src/mctpd.c:
```diff
@@ -268,6 +288,83 @@ __attribute__((format(printf, 1, 2))) static void bug_warn(const char *fmt, ...)
+}
+
+// Check if set is empty
+bool eid_set_is_empty(reserved_eid_set *set)
```
`eid_set_is_empty` isn't used

faizana-nvidia replied:
removed the logic

mkj commented on src/mctpd.c:
```diff
@@ -268,6 +288,83 @@ __attribute__((format(printf, 1, 2))) static void bug_warn(const char *fmt, ...)
 }
 
+// Initialize the set
+void eid_set_init(reserved_eid_set *set)
```
These should be static functions

mkj commented on src/mctpd.c:
```diff
@@ -1690,20 +1802,69 @@ static int endpoint_assign_eid(struct ctx *ctx, sd_bus_error *berr,
 			if (n->peers[e])
 				continue;
+			// todo: check reservation of eid e from other bridge
+			if (check_bridge) {
```
Maybe this should be moved out to a helper function, is getting fairly long and nested. (+- Jeremy's query about `check_bridge`)

faizana-nvidia replied:
updated in above Jeremy's comment

mkj commented on src/mctpd.c:
```diff
@@ -3716,6 +4023,16 @@ static int add_net(struct ctx *ctx, uint32_t net_id)
 	emit_net_added(ctx, net);
+
+	//Initialize reserved eid set
+	warnx("add new netspace for reserved eids for net %d", net->net);
```
This debug warning can be removed

mkj commented on src/mctpd.c:
```diff
@@ -4005,6 +4332,125 @@ static void free_config(struct ctx *ctx)
+		     e < peer->pool_start + peer->pool_size; e++) {
+			if (n->resv_eid) {
+				eid_set_add(n->resv_eid, e);
+				warnx("adding to reserve for net %d eid %d",
```
This is info rather than a warning?

mkj reviewed (commented):
A few style comments

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -4100,6 +4113,114 @@ static void free_config(struct ctx *ctx)
+		}
+	} else if (stat == 0x1) {
+		warnx("%s Allocation was rejected as it was Allocated by other bus \n",
+		      __func__);
```
Sure then we'll simply dont go with ALLOCATE EID for such conflicts, only bridge itself would be populated to dbus object with no pool size

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -51,6 +51,7 @@
 #define OPENBMC_IFACE_COMMON_UUID "xyz.openbmc_project.Common.UUID"
 #define CC_MCTP_DBUS_IFACE_INTERFACE "au.com.codeconstruct.MCTP.Interface1"
 #define CC_MCTP_DBUS_NETWORK_INTERFACE "au.com.codeconstruct.MCTP.Network1"
+#define ERROR_BRIDGE_START_EID 0
```
Ack

jk-ozlabs commented:
Looking pretty good with the changes. Let me know when you're at a stable point and want me to re-review.

faizana-nvidia commented:
> Looking pretty good with the changes. Let me know when you're at a stable point and want me to re-review.

Hello Jeremy/Matt
Thank you for your review comments. Please help in re-review of the MR, mostly all the comments were addressed.

faizana-nvidia commented:
> Neat! I shall do. I've got a couple of other things pending at the moment, so will get on to a proper review shortly, but in the meantime there are a still a couple of things from the earlier reviews:
> 
> * the dbus API changes need to be documented in mctpd.md, using the per-interface formats there

Ack, let me get onto it.
> * the tests need to cover further cases around invalid endpoint allocation requests, and allocation failures

Thanks, I can incorporate the test case.

faizana-nvidia commented on conf/mctpd.conf:
```diff
@@ -5,6 +5,10 @@ mode = "bus-owner"
 [mctp]
 message_timeout_ms = 30
 
+# bus-owner configuration
+[bus-owner]
+max_pool_size = 15
+
```
right, overlooked this part, my bad will update

jk-ozlabs replied:
(I have also added some documentation on the configuration in that branch too)

faizana-nvidia replied:
Ack, reworked my change based on your reference

jk-ozlabs commented on tests/test_mctpd.py:
```diff
@@ -831,3 +831,42 @@ async def test_interface_rename_with_peers(dbus, mctpd):
+        #check if the downstream endpoint eid is contiguous to the bridge endpoint eid
+        assert (eid + i + 1) == br_ep.eid
+        (path, new) = await net.call_learn_endpoint(br_ep.eid)
+        assert path == f'/au/com/codeconstruct/mctp1/networks/1/endpoints/{br_ep.eid}'
```
Same here; no need to check the paths, just that LearnEndpoint succeeded.

faizana-nvidia replied:
Ack

jk-ozlabs commented on src/mctpd.c:
```diff
@@ -3047,6 +3058,33 @@ static int bus_endpoint_get_prop(sd_bus *bus, const char *path,
+		rc = sd_bus_message_append(reply, "y", peer->pool_size);
+	} else if (strcmp(property, "PoolEnd") == 0) {
+		uint8_t pool_end =
+			peer->pool_size ?
```
we know that pool_size > 0, right? otherwise we wouldn't have this interface?

jk-ozlabs commented on src/mctpd.c:
```diff
@@ -3246,6 +3284,26 @@ static const sd_bus_vtable bus_endpoint_cc_vtable[] = {
+			"y",
+			bus_bridge_get_prop,
+			0,
+			SD_BUS_VTABLE_PROPERTY_CONST),
```
Do we need both size and end?

faizana-nvidia replied:
We can keep end and remove size for better readability of pool eids (this would. give range)

jk-ozlabs reviewed (commented):
Looking good, a few things inline.

Since github doesn't support review comments on the commit messages, just a few general pointers for those. Can you keep consistency with the existing messages? Mainly wrapping at 80 cols, and typically sentence format rather than dot-points.

No need to explain individual code-level changes, we can read that from the diff. Instead, general context and intent is better.

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -1691,11 +1719,50 @@ static int endpoint_assign_eid(struct ctx *ctx, sd_bus_error *berr,
+
+			// check for max sized pool from e + 1
+			if (assign_bridge) {
+				int next_pool_start = get_next_pool_start(
+					e, n, ctx->max_pool_size);
+				if (next_pool_start < 0) {
+					warnx("Ran out of EIDs from net %d while"
+					      "allocating bridge downstream endpoint at %s ",
+					      net, dest_phys_tostr(dest));
+					is_pool_possible = false;
+					/*ran out of pool eid : set only bridge eid then
+					find first available bridge eid which is not part of any pool*/
+					for (e = eid_alloc_min;
+					     e <= eid_alloc_max; e++) {
+						if (n->peers[e]) {
+							// used peer may be a bridge, skip its eid range
+							e += n->peers[e]
+								     ->pool_size;
+							continue;
+						}
+						break;
+					}
+				} else if (next_pool_start != e + 1) {
+					// e doesn't have any contiguous max pool size eids available
+					e += next_pool_start;
+					continue;
+				} else {
+					// found contigous eids of max_pool_size from bridge_eid
+					is_pool_possible = true;
+				}
```
okay, I'll update this.

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -1660,15 +1666,37 @@ static int peer_set_mtu(struct ctx *ctx, struct peer *peer, uint32_t mtu)
+		return -EADDRNOTAVAIL;
+	}
+	for (mctp_eid_t e = bridge_eid + 1; e <= bridge_eid + max_pool_size;
+	     e++) {
```
hmm I see what you mean. I'll update

faizana-nvidia commented on tests/test_mctpd.py:
```diff
@@ -831,3 +831,42 @@ async def test_interface_rename_with_peers(dbus, mctpd):
+    net = await mctpd_mctp_network_obj(dbus, iface.net)
+    for i in range(pool_size):
+        br_ep = ep.bridged_eps[i]
+        #check if the downstream endpoint eid is contiguous to the bridge endpoint eid
+        assert (eid + i + 1) == br_ep.eid
```
we can have allocated_pool_size representing pool response from Allocate Endpoint ID and later I can check downstream eid could fall within bridge's eid and its allocated pool sized next eids

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -4138,6 +4165,128 @@ static void free_config(struct ctx *ctx)
+	if (rc) {
+		//reset peer pool
+		peer->pool_size = 0;
+		peer->pool_start = 0;
```
Ack

faizana-nvidia replied:
lol, nowww I understood what you mean. else is pointless here.

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -4138,6 +4165,128 @@ static void free_config(struct ctx *ctx)
+					       peer->phys.ifindex, &gw_addr,
+					       peer->mtu);
+			if (rc < 0) {
+				warnx("Failed to add Gateway route for EID %d: %s",
```
will update

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -4138,6 +4165,128 @@ static void free_config(struct ctx *ctx)
+
+	iid = mctp_next_iid(peer->ctx);
+	req.ctrl_hdr.rq_dgram_inst = RQDI_REQ | iid;
+	req.ctrl_hdr.command_code = MCTP_CTRL_CMD_ALLOCATE_ENDPOINT_IDS;
```
Ack will use

faizana-nvidia commented on tests/test_mctpd.py:
```diff
@@ -831,3 +831,42 @@ async def test_interface_rename_with_peers(dbus, mctpd):
+    assert new
+    # Assert for assigned bridge endpoint ID
+    assert path == f'/au/com/codeconstruct/mctp1/networks/1/endpoints/{eid}'
+    assert new
```
will remove

faizana-nvidia commented:
> * the tests need to cover further cases around invalid endpoint allocation requests, and allocation failures

for this case, we can't really test allocation failure since dbus method AssignEndpoint will at least assign Bridge's own eid, even with case allocated was rejected/failed and we reply bridge's eid path.

Also there are other complications involved while addressing allocation rejection such as
1. bridge needs to already have some eid, some pool size and downstream eids assigned.
2. we have not added case where if endpoint have already been assigned any eid should return same as response to  SET_ENDPOINT_ID command handling.
3. coming back to same even all this happen and we get a rejected case, we can't infer that in our test as we have no validation factor in response of dbus method.

for successful allocation we have a factor like assigned eid to downstream endpoints should be within pool range of bridge and contiguous.

I can add invalid endpoint test though i.e if asked EID for some endpoint is falling within any other MCTP bridge pool.

jk-ozlabs commented:
> for this case, we can't really test allocation failure since dbus method AssignEndpoint will at least assign Bridge's own eid

That's fine. Just check the resulting pool range of the Allocate Endpoint IDs command, which you would have stored on the Endpoint.

It's fine to leave the more stateful tests (ie, where the bridge already has some EID/pool assignment) for later, but there are definitely some low-hanging fruit to address:

 - where the bridge pool request (in the Set Endpoint ID response) is larger than mctpd's max
 - where the bridge pool request would conflict with some other EID, so there is no valid pool available (use the new `dynamic_eid_range` support to help trigger that case, and maybe create a static assignment for a conflicting EID)
 - where the endpoint returns a failure from the Assign Endpoint IDs command (is the pool allocation left in a consistent state?)
 - after a successful bridge pool assignment, where a new non-bridge endpoint is enumerated, that it does not conflict with the new pool

jk-ozlabs commented:
You may want to rebase to current `main` - this means you can drop the `max_pool_size` change, because it has already been merged.

jk-ozlabs commented:
> > *  where the endpoint returns a failure from the Assign Endpoint IDs command (is the pool allocation left in a consistent state?)
>
> For this use-case we remove the peer altogether so will be the pool. peer would never be setup

But you have already sent a Set Endpoint ID, so the peer *is* set up.

I would think that the peer needs to persist, just with no allocated pool range. You could do the equivalent of a `Remove`, but then you would have lost connectivity with the bridge. I think it is reasonable for a failed Allocate Endpoint IDs to only fail the pool setup, the bridge can remain.

Regardless, this is still a reasonable scenario to test.

(i have a review in progress, will send that shortly)

faizana-nvidia commented:
> > > * where the endpoint returns a failure from the Assign Endpoint IDs command (is the pool allocation left in a consistent state?)
> > 
> > 
> > For this use-case we remove the peer altogether so will be the pool. peer would never be setup
> 
> But you have already sent a Set Endpoint ID, so the peer _is_ set up.
> 

I see, I might have misunderstood you, so ask is about failure in Allocate Endpoint ID control command but Bridge's own eid has been assigned. So that should remain (we are on same page here, changes are dong the same). What I meant earlier was with failure of setup endpoint itself will lead to removal of peer (already part of design)

> I would think that the peer needs to persist, just with no allocated pool range. You could do the equivalent of a `Remove`, but then you would have lost connectivity with the bridge. I think it is reasonable for a failed Allocate Endpoint IDs to only fail the pool setup, the bridge can remain.
> 
> Regardless, this is still a reasonable scenario to test.
> 
> (i have a review in progress, will send that shortly)

jk-ozlabs commented:
> I see, I might have misunderstood you, so ask is about failure in Allocate Endpoint ID control command

Yes. Sorry, my fault with the ambiguous wording there: should have been "failure from Allocate Endpoint IDs" :)

jk-ozlabs commented:
> Need one suggestion here too, since we publish the peer right after successfull Setup endpoint id, so by this time we have one assumed pool size of max_pool_size (toml config), this will create new interface `au.com.codeconstruct.MCTP.Bridge1 ` and later we are resetting the pool (meaning pool size =0 pool start 0) due to failure in Allocate Endpoint ID but this would still mean interface is present, should we manually remove Interface au.com.codeconstruct.MCTP.Bridge1?

Hm, good question!

I would say that the important thing is that external consumers of the (Bridge1) interface are able to use the interface functionality immediately, and that the data presented is correct.

So, I think that the correct approach would be to publish the Bridge1 interface once the pool assignment is finalised. This would mean waiting until the Allocate Endpoint IDs command has succeeded, and after the gateway range route has been added.

faizana-nvidia commented:
Hello Jeremy,

Thank you taking out time in the reviews, all comments have been addressed, open to any further modifications needed.

jk-ozlabs commented:
OK cool. I'll take a look through soon.

jk-ozlabs commented on tests/test_mctpd.py:
```diff
@@ -971,3 +972,125 @@ async def test_assign_dynamic_bridge_eid(dbus, mctpd):
+    # Interface should not be present for failed pool allocation
+    with pytest.raises(asyncdbus.errors.InterfaceNotFoundError):
+        bridge_obj = await dbus.get_proxy_object(MCTPD_C, path)
+        await bridge_obj.get_interface(MCTPD_ENDPOINT_BRIDGE_I)
```
Invalid whitespace here?

faizana-nvidia replied:
ack removed

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -4352,6 +4395,18 @@ static int endpoint_allocate_eid(struct peer *peer)
+			if (rc) {
+				warnx("Failed to emit add %s signal for endpoint %d : %s",
+				      CC_MCTP_DBUS_IFACE_BRIDGE, peer->eid,
+				      strerror(-rc));
```
Ack : if check needs to change (on success positive integer is expected) if (rc < 0)

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -1673,11 +1738,34 @@ static int endpoint_assign_eid(struct ctx *ctx, sd_bus_error *berr,
 		/* Find an unused dynamic EID */
 		for (e = ctx->dyn_eid_min; e <= ctx->dyn_eid_max; e++) {
-			if (n->peers[e])
+			if (n->peers[e]) {
+				// used peer may be a bridge, skip its eid range
+				e += n->peers[e]->pool_size;
 				continue;
+			}
+
+			// check for max sized pool from e + 1
+			if (assign_bridge) {
+				bool out_of_eids = false;
+				e = get_bridge_eid(n, ctx, &is_pool_possible,
+						   &out_of_eids, e, net, dest);
+				if (!is_pool_possible) {
+					if (!out_of_eids)
+						continue;
+					else {
+						if (e > ctx->dyn_eid_max)
+							break;
+					}
+				}
+			}
+
 			rc = add_peer(ctx, dest, e, net, &peer);
 			if (rc < 0)
 				return rc;
+			if (assign_bridge && is_pool_possible) {
+				peer->pool_size = ctx->max_pool_size;
+				peer->pool_start = e + 1;
+			}
```
Ack, I'll pull in this commit on very top if thats okay?

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -2224,10 +2342,22 @@ static int method_assign_endpoint_static(sd_bus_message *call, void *data,
+				bug_warn("%s: Bad old net %d", __func__, netid);
+				return -EPROTO;
+			}
+			if (is_eid_in_bridge_pool(n, ctx, eid)) {
```
It would, I'll update

jk-ozlabs replied:
> LearnEndpoint method under `au.com.codeconstruct.MCTP.Network1` interface was purposed to be used for already routed endpoints which in a way could also be used for bridged endpoints as stated in the `mctpd.md`

I was more referring to BusOwner1.LearnEndpoint here, which is another facility (alongside LearnEndpointStatic) which may consume an eid without the bridge pool checking, as it calls `add_peer()` directly.

Ideally, we would want to check *all* EID allocations via add_peer(). That would involve allowing EID assignments that have addresses that we "exepect": if they're routed behind some other peer that we know about, it should use an EID in that pool's range. If not, we should reject the assignment.

As you say though, we may be missing some information to properly do that. We can defer this checking to later if necessary.

However, In the case of BusOwner1.LearnEndpoint though, we *know* that this is a directly-connected peer (we are addressing it physically), so should probably disallow any learnt EID in a pool range in that path too.

faizana-nvidia replied:
> This may break an intended usage of `Network1.LearnEndpoint`, no? where:
> 
> * a bridge is managed by mctpd
> * a user calls `Network1.LearnEndpoint` on a peer behind that bridge

Yes this check under ```add_peer()``` would break the intended Network1.LearnEndpoint only for bridged eids, I wanted to avoid the need to manipulate add_peer() signature right now and may be address this better on next MCTP Bridge Poll PR, but I agree on ``` make sure we don't "paint ourselves into a corner" though, where a future change would break existing supported usage.``` so I'll update this

With this at least our intention  ```we only support bridging through mctpd-managed bridges``` would be taken care of (hopefully) but as you said too, we might need some netlink data for (non mctpd managed) bridged routes.

Will update the ```mctpd.md``` document about managing only MCTPD managed bridges

faizana-nvidia replied:
updated the patch, to avoid check altogether when Network1.LearnEndpoint is called, so that we don't break things for it for now.

jk-ozlabs reviewed (commented):
Looking good. I have some comments on some minor things, and would like to do bit of experimentation on:

 - whether we can make Network1.LearnEndpoint more accurately-selective on the bridge states
 - adding further test cases

I will follow up with the outcomes of those shortly; likely as proposed additions to the series.

faizana-nvidia commented on tests/mctpenv/__init__.py:
```diff
@@ -303,6 +303,9 @@ def __init__(self, iface, lladdr, ep_uuid = None, eid = 0, types = None):
         self.eid = eid
         self.types = types or [0]
         self.bridged_eps = []
+        self.pool_size = 0
+        self.allocated_pool = None # [start, size]
```
Since its an MCTP Bridge, I'm assuming pool_size could represent all downstream devices which are connected to the Bridge Endpoint (representing physical connection to the Bridge), then during SET_ENDPOINT_ID command it would respond as pool size to be asked to BO.

Thus two names pool_size and allocated_pool [start, size].

faizana-nvidia replied:
may be for simplifying this I could rename it as dynamic_pool_size from pool_size

jk-ozlabs replied:
that's not really a simplification.

At present, the only value you're storing in `pool_size` is `len(self.bridged_eps)`. We don't need another copy of that value.

I will propose a patch that adjusts these tests.

jk-ozlabs commented on tests/test_mctpd.py:
```diff
@@ -926,3 +926,42 @@ async def test_config_dyn_eid_range_max(nursery, dbus, sysnet):
+    with pytest.raises(asyncdbus.errors.DBusError) as ex:
+        await mctp.call_assign_endpoint_static(dev2.lladdr, ep.eid + 1)
+
+    assert str(ex.value) == "Request failed"
```
The last line of the file has no newline; the github UI is marking it with a warning symbol.

faizana-nvidia commented on tests/mctpenv/__init__.py:
```diff
@@ -367,6 +372,23 @@ async def handle_mctp_control(self, sock, addr, data):
+                    alloc_status = 0x01
+                else:
+                    num_eps = min(pool_size, self.pool_size)
+                    self.allocated_pool = [pool_start, num_eps]
```
Ack, I'll add your suggested changes

faizana-nvidia commented on src/mctpd.c:
```diff
@@ -4360,6 +4386,128 @@ static void free_config(struct ctx *ctx)
+			rc = -1;
+			goto out;
+		}
+		*allocated_pool_size = resp->eid_pool_size;
+		*allocated_pool_start = resp->eid_set;
```
okay, sure

faizana-nvidia replied:
>I think this is resolved, now that send_allocate_endpoint_id will only set those to the provided sizes?

I suppose its correct what you are saying, since peer->pool_size and peer->pool_start is the asked valued going as argument to send_allocate_endpoint_id() and once it has been accepted it would remain the same value, had it been any other value we would have bailed out as an error, so no sense in even storing this as allocated_pool_start and allocated_pool_size

jk-ozlabs replied:
I'm still happy to store the values (for example, if we ever do implement adjustments to allow for smaller pool ranges). Just that it's (very slightly) less prone to future errors if you use the `eid_pool_size` and `eid_pool_start` values instead of the values directly from `resp`.

jk-ozlabs commented:
I am keeping my pending changes in a `review/bridge` branch: https://github.com/CodeConstruct/mctp/compare/main...jk-ozlabs:mctp:review/bridge , if that helps. I will be updating during review though.

If you want to squash any of those into your series, feel free, just let me know first.

faizana-nvidia commented:
> I am keeping my pending changes in a `review/bridge` branch: [main...jk-ozlabs:mctp:review/bridge](https://github.com/CodeConstruct/mctp/compare/main...jk-ozlabs:mctp:review/bridge) , if that helps. I will be updating during review though.
> 
> If you want to squash any of those into your series, feel free, just let me know first.

Thank you for the commits.
I'm gonna pick last two (splitting of test case for static bridge.. and netlearn test case) rest I'm gonna update directly.

faizana-nvidia commented:
Thank you for rebasing the PR with your br :)
> OK, thanks for taking a look!
> 
> My plan is to push to the branch for this PR, and then merge. That way it's your original PR that gets tracked as the merge point for this feature.

Ack, would be better this way.

jk-ozlabs commented:
No problem - we needed the rebase in order to resolve conflicts with 9711f0bf1b8f47d4d6e3ac58363a5f35ac8c4bd7.

Next step is to do a little testing on hardware we have here, then we should be good to merge. Let me know (soon!) if you have any further alterations beforehand.
